python/orca/tutorial/pytorch/NCF/train_data_loader.py

```diff
@@ -0,0 +1,237 @@
+from sklearn.model_selection import train_test_split
+
+def load_all():
+    """ We load all the files here to save time in each epoch. """
+    data_X = pd.read_csv(
+        args.dataset+"/ratings.dat", 
+        sep="::", header=None, names=['user', 'item'], 
+        usecols=[0, 1], dtype={0: np.int32, 1: np.int32})
+    
+    user_num = data_X['user'].max() + 1
+    item_num = data_X['item'].max() + 1
+
+    data_X = data_X.values.tolist()
+
+    # load ratings as a dok matrix
+    train_mat = sp.dok_matrix((user_num, item_num), dtype=np.float32)
+    for x in data_X:
+        train_mat[x[0], x[1]] = 1.0
+        
+    train_data, test_data = train_test_split(data_X, test_size=0.1, random_state=100)
+    
+    return train_data, test_data, user_num, item_num, train_mat
+
+class NCFData(data.Dataset):
+    def __init__(self, features, 
+                num_item, train_mat=None, num_ng=0, is_training=None):
+        super(NCFData, self).__init__()
+        """ Note that the labels are only useful when training, we thus 
+            add them in the ng_sample() function.
+        """
+        self.features_ps = features
+        self.num_item = num_item
+        self.train_mat = train_mat
+        self.num_ng = num_ng
+        self.is_training = is_training
+        self.labels = [0 for _ in range(len(features))]
+
+    def ng_sample(self):
+        assert self.is_training, 'no need to sampling when testing'
+
+        self.features_ng = []
+        for x in self.features_ps:
+            u = x[0]
+            for t in range(self.num_ng):
+                j = np.random.randint(self.num_item)
+                while (u, j) in self.train_mat:
+                    j = np.random.randint(self.num_item)
+                self.features_ng.append([u, j])
+
+        labels_ps = [1 for _ in range(len(self.features_ps))]
+        labels_ng = [0 for _ in range(len(self.features_ng))]
+
+        self.features_fill = self.features_ps + self.features_ng
+        self.labels_fill = labels_ps + labels_ng
+
+    def __len__(self):
+        return (self.num_ng + 1) * len(self.labels)
+    
+    def __getitem__(self, idx):
+        features =  self.features_fill if self.is_training else self.features_ps
+        labels =  self.labels_fill if self.is_training else self.labels
+
+        user = features[idx][0]
+        item = features[idx][1]
+        label = float(labels[idx])
+        return [user, item] ,label
+
+# prepare the train and test datasets
+train_data, test_data, args.user_num, args.item_num, train_mat = load_all()
+
+# construct the train and test dataloader
+train_dataset = NCFData(
+        train_data, args.item_num, train_mat,args.num_ng, True)
+test_dataset = NCFData(
+        test_data, args.item_num, train_mat, 0, False)
+train_loader = data.DataLoader(train_dataset,
+        batch_size=args.batch_size, shuffle=True, num_workers=0)
+test_loader = data.DataLoader(test_dataset,
+        batch_size=args.test_num_ng+1, shuffle=False, num_workers=0)

```

Comment:
do not create the dataset and dataloader on the driver, put them separately in train_loader_func and test_loader_func